### PR TITLE
Feature: add application scopes framework and imp health scope

### LINKED ISSUE #9: Support scopes
The specification defines several scopes. None of them are present or provided by the current implementation.

### PR DESCRIPTION
**Pull request by wonderflow**
This PR is mainly to add application scope framework and health scope

This PR lasts several weeks mostly because the spec is changing.

https://github.com/microsoft/hydra-spec/issues/130
https://github.com/microsoft/hydra-spec/issues/144
https://github.com/microsoft/hydra-spec/pull/133
https://github.com/microsoft/hydra-spec/pull/127 
https://github.com/microsoft/hydra-spec/pull/177
https://github.com/microsoft/hydra-spec/pull/196

## add application scopes framework 

- [x] 1. add framework for network/health
- [x] 2. create health scope instance
- [x] 3. check component scope overlap
- [x] 4. add/remove Component to scope

## add health scope

- [x] 1. create a sub project to implement healthscope
- [x] 2. periodically aggregate status of health scope
- [x] 3. create a service to make health scope available for others
- [x] 4. docs for how to use it.

This is same with #160 as we have changed repo, I have to make a new PR.

fixes #9

### CODE REVIEW

**Comment by wonderflow** [comment]
It seems the version of openssl installed by base image is 1.1.0 which didn't match the requirement of the updated lib.

I'm trying to figure out whether I should use a new base image

**Comment by wonderflow** [comment]
@technosophos @hongchaodeng  I think this PR is ready for review now.

I have reset all the commits and split them into four big commits. Each commit is a different part of code so you can review them one by one.

File: Dockerfile [hunk 1]
```diff
@@ -1,27 +1,86 @@
...
+&& apt-mark auto '.*' > /dev/null \
+&& { [ -z "$savedAptMark" ] || apt-mark manual $savedAptMark; } \
+&& find /usr/local -type f -executable -exec ldd '{}' ';' \
+		| awk '/=>/ { print $(NF-1) }' \
+		| sort -u \
+		| xargs -r dpkg-query --search \
+		| cut -d: -f1 \
+		| sort -u \
+		| xargs -r apt-mark manual \
+&& apt-get purge -y --auto-remove -o APT::AutoRemove::RecommendsImportant=false
```
Comment (wonderflow) [inline Dockerfile]: kube-rs 0.6.1 need openssl 1.1.1 while using apt-get in debian:stretch-slim didn't work, so I have to compile from source here.

File: src/schematic/scopes/network.rs [hunk 2]
```diff
@@ -0,0 +1,117 @@
+use crate::schematic::configuration::ComponentConfiguration;
+/// Network scope is defined as https://github.com/oam-dev/spec/blob/master/4.application_scopes.md#network-scope
+/// Now we don't really implement network scope, this is just a framework as the spec describe.
```
Comment (hongchaodeng) [inline src/schematic/scopes/network.rs]: add _TODO_

File: Dockerfile (hunk 1 above)
Comment (hongchaodeng) [reply Dockerfile]: Maybe add the comments in the Dockerfile?

File: healthscope/README.md [hunk 3]
```diff
@@ -0,0 +1,197 @@
+# Health Scope Controller
+
+Health Scope Controller used for periodically check health scope crd and check the health of all the components related.
```
Comment (hongchaodeng) [inline healthscope/README.md]:
```suggestion
HealthScope controller is used to periodically check the health of all the components and update health scope custom resources.
```

File: healthscope/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+            match client.request::<ObjectList<HealthScopeObject>>(req) {
+                Ok(health_scopes) => {
+                    for scope in health_scopes.items {
+                        if let Err(res) = aggregate_health(&client, scope, ns.clone()) {
+                            // Log the error and continue.
+                            error!("Error processing event: {:?}", res)
+                        };
+                    }
+                }
+                Err(e) => log::error!("get health scope list err {}", e),
```
Comment (hongchaodeng) [inline healthscope/src/main.rs]:
```suggestion
                Err(e) => error!("get health scope list err {}", e),
```

File: healthscope/README.md [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+
+## What will health scope controller do?
+
+1. periodically check all component health and update the CR status.
```
Comment (hongchaodeng) [inline healthscope/README.md]:
```suggestion
1. periodically check health status of components and update the HealthScope resource status.
```

File: healthscope/README.md [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+2. serve as a http server, to output aggregated health information.
+
+## How to install?
+
+You could use helm to install it:
```
Comment (hongchaodeng) [inline healthscope/README.md]:
```suggestion
Use helm to install:
```

File: healthscope/README.md [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+
+```shell script
```
Comment (hongchaodeng) [inline healthscope/README.md]:
```suggestion
```shell
```

File: healthscope/README.md [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+helm install healthscope ./charts/healthscope
+```
+
+## How to use?
+
+After healthscope was installed by helm charts, you could setup an endpoint to it.
```
Comment (hongchaodeng) [inline healthscope/README.md]:
```suggestion
After healthscope controller is installed by helm charts, setup an endpoint to it.
```

File: healthscope/README.md [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+By default the charts will use ClusterPort, you could use port mapper to make health scope accessible.
```
Comment (hongchaodeng) [inline healthscope/README.md]:
```suggestion
By default the charts will use ClusterPort, we can use port-forward to access healthscope controller.
```

File: healthscope/README.md [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+
+```shell script
+export POD_NAME=$(kubectl get pods -l "app.kubernetes.io/name=healthscope,app.kubernetes.io/instance=health" -o jsonpath="{.items[0].metadata.name}")
+kubectl port-forward $POD_NAME 8080:80
+```
+
+Then you'll be able to visit http://127.0.0.1:8080 to get health scope status.
```
Comment (hongchaodeng) [inline healthscope/README.md]:
```suggestion
Then we'll be able to visit http://127.0.0.1:8080 to query status of healthscope resource.
```

File: healthscope/src/main.rs [hunk 11]
```diff
@@ -0,0 +1,363 @@
...
+            }
+            cnt = (cnt + 1) % 10;
```
Comment (hongchaodeng) [inline healthscope/src/main.rs]: remove debugging code?

File: healthscope/src/main.rs [hunk 12]
```diff
@@ -0,0 +1,363 @@
...
+            if cnt == 0 {
+                debug!("health scope aggregate loop running...");
+            }
+            //FIXME: we could change this to use an informer if we have a runtime controller queue
+            std::thread::sleep(std::time::Duration::from_secs(1));
```
Comment (hongchaodeng) [inline healthscope/src/main.rs]: 1s as default might be too frequent. Maybe 5s?

File: healthscope/src/main.rs [hunk 13]
```diff
@@ -0,0 +1,363 @@
...
+        );
+    })
+    .join()
+    .unwrap();
+
+    server.join().unwrap();
+    health_scope_watch.join().unwrap()
+}
+
+use std::{
```
Comment (hongchaodeng) [inline healthscope/src/main.rs]: Put this in the head of file?

File: healthscope/src/main.rs [hunk 14]
```diff
@@ -0,0 +1,363 @@
...
+                task.notify()
+            }
+        });
+
+        HealthFuture { shared_state }
+    }
+}
+
+type BoxFut = Box<dyn Future<Item = Response<Body>, Error = hyper::Error> + Send>;
+fn serve_health(req: Request<Body>) -> BoxFut {
```
Comment (hongchaodeng) [inline healthscope/src/main.rs]: add some comments?

File: healthscope/src/main.rs [hunk 15]
```diff
@@ -0,0 +1,363 @@
...
+            let instance = path.trim_start_matches('/').to_string();
+            info!("{} health scope requested", instance);
+            return Box::new(HealthFuture::new(instance));
+        }
+        _ => *response.status_mut() = StatusCode::NOT_FOUND,
+    }
+    Box::new(future::ok(response))
+}
+
+fn request_health(instance_name: String) -> Result<String, Error> {
```
Comment (hongchaodeng) [inline healthscope/src/main.rs]: add some comments?

File: healthscope/src/main.rs [hunk 16]
```diff
@@ -0,0 +1,363 @@
...
+                        health = "unhealthy"
+                    }
+                };
+            })
+        })
+    });
+    Ok(health.to_string())
+}
+
+fn aggregate_health(
```
Comment (hongchaodeng) [inline healthscope/src/main.rs]: aggregate_health -> aggregate_components_health

**Comment by hongchaodeng** [comment]
ref https://github.com/oam-dev/spec/pull/156

File: healthscope/README.md (hunk 7 above)
Comment (wonderflow) [reply healthscope/README.md]: my IDE auto complete `shell` to `shell script `

File: healthscope/src/main.rs (hunk 12 above)
Comment (wonderflow) [reply healthscope/src/main.rs]: ok

**Comment by wonderflow** [comment]
@hongchaodeng  thanks very much, all comments are fixed in the last commit

**Comment by fibonacci1729** [comment]
@wonderflow i'll review this today.

File: healthscope/src/main.rs [hunk 17]
```diff
@@ -0,0 +1,368 @@
...
+                shared_state.resp.clone(),
+            ))))
+        } else {
+            shared_state.task = Some(current());
+            Ok(Async::NotReady)
+        }
+    }
+}
+
+// FIXME kube-rs client doesn't support async call so we have to wrap it in HealthFuture here.
```
Comment (technosophos) [inline healthscope/src/main.rs]: This looks like a solid solution while we wait for the Kube upstream.

**Comment by technosophos** [comment]
The rebase that needs to be done is really simple, so feel free to merge as soon as you have rebased. I am excited to see this new addition!

**Comment by wonderflow** [comment]
Thanks @technosophos , I've rebased.

I noticed changes from https://github.com/oam-dev/rudr/pull/375 has been overrided by #366. I rebased it back here.

**Comment by wonderflow** [comment]
CI has passed, merging

**Pull request merged**